skimage/segmentation/fast_draw.py

```diff
@@ -0,0 +1,336 @@
+        
+        assert (beta > 0), 'beta should be positive.'
+        self.beta = beta
+        self.return_full_prob = return_full_prob
+        self.image = rgb2grey(image)
+        ## It is important to normalize the data between [0,1] so that beta
+        ## makes sens
+        self.image = img_as_float(self.image)
+        
+        if downsampled_size > min(self.image.shape):
+            warn('The size of the downsampled image if larger than the '
+                 'original image. The computation time could be affected.')
+        
+        ## Compute the graph's Laplacian for both the full resolution `L` 
+        ## and the down-sampled `ds_L` images
+        self.L = self._buildGraph(self.image)
+        ratio = float(self.image.shape[0])/self.image.shape[1]
+        self.dim = (int(downsampled_size*ratio), downsampled_size)
+        self.ds_image = resize(self.image, self.dim, order=0,
+                                 preserve_range=True)
+        self.ds_L = self._buildGraph(self.ds_image)
+        
+        ## Initialize the ROI to zero
+        self.ds_maskROI = np.zeros_like(self.ds_image, dtype=np.bool)
+        ## `full_to_ds_ratio` is used to convert labels from full resolution
+        ## image to down-sampled image
+        self.full_to_ds_ratio = (float(self.dim[0])/self.image.shape[0],
+                                 float(self.dim[1])/self.image.shape[1])
+
+    def _buildGraph(self, image):
+        """Builds the graph: vertices, edges and weights from the `image` 
+        and returns the graph's Laplacian `L`.
+        """
+        # Building the graph: vertices, edges and weights
+        nsize = reduce(lambda x, y : x*y, image.shape,1)
+        vertices = np.arange(nsize).reshape(image.shape)
+        edges_right = np.vstack((vertices[:, :-1].ravel(),
+                                 vertices[:, 1:].ravel()))
+        edges_down = np.vstack((vertices[:-1].ravel(),
+                                vertices[1:].ravel()))
+        edges = np.hstack((edges_right, edges_down))
+        
+        gr_right = np.abs(image[:, :-1] - image[:, 1:]).ravel()
+        gr_down = np.abs(image[:-1] - image[1:]).ravel()
+        weights = np.exp(-self.beta * np.r_[gr_right, gr_down]**2)+1e-6
+        
+        # Compute the graph's Laplacian L
+        pixel_nb = edges.max() + 1
+        diag = np.arange(pixel_nb)
+        i_indices = np.hstack((edges[0], edges[1]))
+        j_indices = np.hstack((edges[1], edges[0]))
+        data = np.hstack((-weights, -weights))
+        lap = coo_matrix((data, (i_indices, j_indices)),
+                                shape=(pixel_nb, pixel_nb))
+        connect = - np.ravel(lap.sum(axis=1))
+        lap = coo_matrix((np.hstack((data, connect)),
+                         (np.hstack((i_indices, diag)),
+                         np.hstack((j_indices, diag)))),
+                         shape=(pixel_nb, pixel_nb))
+        L = lap.tocsr()
+        
+        return L
```

Comment:
Couldn't you use the function ``_make_laplacian_sparse`` from the ``random_walker_segmentation.py`` module? The less code the better :-).